### PR TITLE
Make new api compatible with the Signing interface

### LINKED ISSUE #1263: Support custom signing implementations in Metadata.sign method
**Description of issue or feature request**:
Allow TUF integrators to easily use custom metadata signing implementations instead of the currently supported securesystemslib one.

An urgent use case for this is the PEP458/Warehouse integration, where metadata must be signed with keys stored in a Hashicorp Vault and the implementation is already available.

NOTE: This feature request only concerns the `tuf.api.Metadata.sign` method and not the legacy signing facilities `write` and `writeall` in repository tool.

**Current behavior**:
`tuf.api.Metadata.sign(key, ...)` calls `securesystemslib.keys.create_signature` with the passed sslib/json-formatted private key.

**Expected behavior**:
`tuf.api.Metadata.sign` takes an abstract signer parameter, which implements signing. A default signer may resolve to the currently used  `securesystemslib.keys.create_signature`.

**Implementation considerations**:
- Different signers require completely different *"keys"*, e.g. the currently used generic implementation requires an sslib/json-formatted private key, whereas signing implementations that don't have direct access to the key may only need a keyid (see GPG or HSM references below). As a consequence the key parameter in `Metadata.sign` should probably be encapsulated within the signer.

- The abstract signer interface must strictly define the format of the returned signature for metadata interoperability.

- The abstract signer interface design should also keep in mind other key-related tasks, such as signature verification, public key export to sslib metadata format, and keyid generation.

- `securesystemslib.keys.create_signature` is already an abstract interface, where the json-formatted private key parameter format is generic enough to hold different key types and to choose one of many supported signing algorithms, identified by the key's `scheme` field. (see public key metadata format reference below). However, it is not suited to extend to custom implementations. A new abstract signer interface should integrate well with the existing securesystemslib infrastructure and/or aim to replace it. It should not become yet another co-existing interface and/or layer of abstraction (see public API overview reference below).

----

**Related work and references**: *(for the very ambitious reader)*

- securesystemslib key overview  (to coordinate this feature request with the existing infrastructure)
  - general public API overview -- [sslib#270](https://github.com/secure-systems-lab/securesystemslib/issues/270)
  - generic signing interface currently used by TUF -- [sslib.keys.create_signature](https://github.com/secure-systems-lab/securesystemslib/blob/v0.18.0/securesystemslib/keys.py#L602)
  - GPG signing interface -- [sslib.gpg.functions.create_signature](https://github.com/secure-systems-lab/securesystemslib/blob/v0.18.0/securesystemslib/gpg/functions.py#L38)
  - WIP: HSM signing interface -- [sslib.hsm.create_signature](https://github.com/secure-systems-lab/securesystemslib/blob/2e20c56c2e65181b38511927d15a2b758f73ddec/securesystemslib/hsm.py#L358)
  - general public key metadata format -- [sslib#308](https://github.com/secure-systems-lab/securesystemslib/issues/308)
  -  in-memory key representation for generic signing interface -- [sslib#310](https://github.com/secure-systems-lab/securesystemslib/issues/310)
  - stand-alone on-disk key format for generic signing interface -- [sslib#309](https://github.com/secure-systems-lab/securesystemslib/issues/309)

- Recently added abstract filesystem interface (for design inspiration) -- https://github.com/theupdateframework/tuf/issues/1009

- @trishankatdatadog's Hashicorp Vault interface POC (as reference for PEP458/warehouse implementation and for design inspiration) -- [VaultKey](https://github.com/theupdateframework/tuf/blob/495e10f346342892b3ec645b3ce488ede64dd30d/tuf/api/keys.py#L167-L429)

- Abstract `pyca/cryptograhpy` key interface with sign method (for design inspiration) -- [RSAPrivateKey](https://github.com/pyca/cryptography/blob/d6535b68455767be37402ad7ce185d2e459daec8/src/cryptography/hazmat/primitives/asymmetric/rsa.py#L15-L44)

- "Historical" external signing API feature request (has some interesting discussion and designs but targets legacy TUF codebase) -- https://github.com/theupdateframework/tuf/issues/864

- in-toto metadata model sign methods (alternative but less flexible approach to support different signing implementations)
  - implementation -- [Metadata.sign*](https://github.com/in-toto/in-toto/blob/v1.0.0/in_toto/models/metadata.py#L137-L195)
  - related design discussion (replace serialization/deserialization with sign/verify!) -- [sslib#272](https://github.com/secure-systems-lab/securesystemslib/issues/272#issuecomment-691056186)

### PR DESCRIPTION
**Pull request by MVrachev**
Related to: #1263

**Description of the changes being introduced by the pull request**:

This pr makes TUF compatible with changes introduced in https://github.com/secure-systems-lab/securesystemslib/pull/319.

**Please wait for the securesystemlib pr to be merged first!**

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

### CODE REVIEW

File: tuf/api/metadata.py [hunk 1]
```diff
@@ -203,7 +214,8 @@ def sign(self, key: JsonDict, append: bool = False) -> JsonDict:
             A securesystemslib-style signature object.
 
         """
-        signature = create_signature(key, self.signed.to_canonical_bytes())
+        signer = SSlibSigner(key)
```
Comment (lukpueh) [inline tuf/api/metadata.py]: I think we will want to make `signer` a parameter that replaces `key`.

**Review by lukpueh — COMMENTED** [review]
Cool stuff, @MVrachev! I only noticed one thing (see inline comment), but we can revisit once secure-systems-lab/securesystemslib#319 has landed.

On a general note, I appreciate slightly more informative commit messages.

**Comment by MVrachev** [comment]
> Cool stuff, @MVrachev! I only noticed one thing (see inline comment), but we can revisit once [secure-systems-lab/securesystemslib#319](https://github.com/secure-systems-lab/securesystemslib/pull/319) has landed.
> 
> On a general note, I appreciate slightly more informative commit messages.

Sorry for the lack of a good commit message, you are right.

I will update this pr when the securesystemslib signer interface is merged and I will add a better commit message before mark it
as a non-draft pr.

**Comment by MVrachev** [comment]
I updated the pr addressing Lukas comment https://github.com/theupdateframework/tuf/pull/1272#discussion_r573718513, adding a new commit fixing one comment in the code, and making the commit message more descriptive.

**Comment by MVrachev** [comment]
Updated the pr after we have bumped securesystemslib version to 0.20.0 which supports the new Signer interface.

**Comment by joshuagl** [comment]
Thanks for this PR Martin, I'm hoping to give it a thorough review ASAP.

One quick observation: we should bump the minimum version of securesystemslib in the `install_requires` entry of setup.py to the version with the `Signer` abstraction.

File: tuf/api/metadata.py [hunk 2]
```diff
@@ -146,8 +152,13 @@ def from_json_file(
     # Serialization.
     def to_dict(self) -> JsonDict:
         """Returns the JSON-serializable dictionary representation of self. """
+
+        signatures_arr = []
```
Comment (joshuagl) [inline tuf/api/metadata.py]: Nit: this variable name seems odd (I think arr implies array, but the data structure is a list). Why not just use `signatures`, like in `from_dict` above?

**Review by joshuagl — CHANGES_REQUESTED** [review]
The small diffstat of this PR feels like a nice validation of the `Signer` interface.

This LGTM, thank you @MVrachev. Could you please add a patch to the PR to update the securesystemslib version in setup.py's `install_requires`? Also, your first patch will no longer be required once we merge #1293.

**Comment by MVrachev** [comment]
I addressed your comments @joshuagl  and fixed the conflicts.

File: tuf/api/metadata.py [hunk 3]
```diff
@@ -92,12 +93,17 @@ class also that has a 'from_dict' factory method. (Currently this is
...
             raise ValueError(f'unrecognized metadata type "{_type}"')
 
+        signatures = []
+        for signature in metadata['signatures']:
+            new_signature = Signature(signature['keyid'], signature['sig'])
+            signatures.append(new_signature)
+
         # NOTE: If Signature becomes a class, we should iterate over
         # metadata['signatures'], call Signature.from_dict for each item, and
         # pass a list of Signature objects to the Metadata constructor intead.
```
Comment (lukpueh) [inline tuf/api/metadata.py]:
```suggestion
```
Please remove this note.

File: tuf/api/metadata.py [hunk 4]
```diff
@@ -92,12 +93,17 @@ class also that has a 'from_dict' factory method. (Currently this is
         else:
             raise ValueError(f'unrecognized metadata type "{_type}"')
 
+        signatures = []
+        for signature in metadata['signatures']:
+            new_signature = Signature(signature['keyid'], signature['sig'])
```
Comment (lukpueh) [inline tuf/api/metadata.py]: Two minor nits:
- I think the data type is a more meaningful distinguishing factor than the instantiation time.
  What do you think about `signature`/`signature_obj`,  or `signature_dict`/`signature`, or `signature_dict`/`signature_obj` (note: I am aware that a dict is also an object, but still). 
- Why not call `Signature.from_dict`? Then we don't have to know here what's going on inside a signature dictionary.

**Review by lukpueh — CHANGES_REQUESTED** [review]
Except for two minor nits this is good to go.

File: tuf/api/metadata.py (hunk 4 above)
Comment (MVrachev) [reply tuf/api/metadata.py]: Both of those remarks make sense. 
I totally forgot that I could use `Signature.from_dict()` and `signature_obj` sounds like a reasonable name.

**Comment by MVrachev** [comment]
I updated according to the last @lukpueh remarks.

File: tuf/api/metadata.py [hunk 5]
```diff
@@ -20,10 +20,11 @@
 
 import tempfile
 
+from securesystemslib.keys import verify_signature
 from securesystemslib.util import persist_temp_file
+from securesystemslib.signer import Signer, Signature
```
Comment (lukpueh) [inline tuf/api/metadata.py:25]: Re-ordered imports alphabetically.

File: tuf/api/metadata.py [hunk 6]
```diff
@@ -90,12 +91,14 @@ def from_dict(cls, metadata: Mapping[str, Any]) -> 'Metadata':
         else:
             raise ValueError(f'unrecognized metadata type "{_type}"')
 
-        # NOTE: If Signature becomes a class, we should iterate over
-        # metadata['signatures'], call Signature.from_dict for each item, and
-        # pass a list of Signature objects to the Metadata constructor instead.
+        signatures = []
+        for signature in metadata.pop('signatures'):
```
Comment (lukpueh) [inline tuf/api/metadata.py:95]: Now popping the `signatures` list for consistently destructing the passed dictionary as described in the newly added "Side Effect" docstring section.

File: tuf/api/metadata.py [hunk 7]
```diff
@@ -178,13 +186,14 @@ def to_file(
...
     def sign(
-        self, key: Mapping[str, Any], append: bool = False,
+        self, signer: Signer, append: bool = False,
         signed_serializer: Optional[SignedSerializer] = None
     ) -> Dict[str, Any]:
         """Creates signature over 'signed' and assigns it to 'signatures'.
 
         Arguments:
-            key: A securesystemslib-style private key object used for signing.
+            signer: An object implementing the securesystemslib.signer.Signer
```
Comment (lukpueh) [inline tuf/api/metadata.py:195]: Did `s/singer/signer` to fix a typo I had missed in my review.

**Review by lukpueh — COMMENTED** [review]
Hey @MVrachev, I took the liberty to rebase your PR on top of #1279, fixing conflicts and making minimal changes. See comments inline...

To get a full diff between your old diff and my new rebased diff you can do:
```
git range-diff develop 3050fb6 49aa0fc
```

I think we can merge. But maybe should consult with a 3rd-party, now that we are both authors? (ping @joshuagl, @jku, @sechkova)

**Comment by MVrachev** [comment]
I agree with your changes @lukpueh.
Let us wait and see what the others will say.

**Review by joshuagl — APPROVED** [review]
LGTM. Thanks for the changes @MVrachev and the rebase/fixes @lukpueh

**Review by lukpueh — APPROVED** [review]

**Comment by lukpueh** [comment]
Ping @woodruffw! `Metadata.sign` now accepts a custom `Signer`. :)

**Comment by trishankatdatadog** [comment]
Very cool, thanks for all the great work, guys!

Will, you should be able to reuse a lot of the Vault-specific code I wrote as part of [this PR](https://github.com/theupdateframework/tuf/pull/1060)...

**Pull request merged**